File: app/src/main/java/tipz/browservio/settings/MaterialSwitchPreference.java
package tipz.browservio.settings;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.TypedArray;
import android.util.AttributeSet;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.preference.SwitchPreferenceCompat;

import tipz.browservio.Application;
import tipz.browservio.R;
import tipz.browservio.utils.CommonUtils;

public class MaterialSwitchPreference extends SwitchPreferenceCompat {
    private final SharedPreferences pref;
    private final String mPreferenceTag;
    private final boolean mNeedReload;

    public MaterialSwitchPreference(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);

        pref = ((Application) getContext().getApplicationContext()).pref;

        // Get attrs
        TypedArray a = getContext().obtainStyledAttributes(attrs, R.styleable.MaterialSwitchPreference);
        mPreferenceTag = a.getString(R.styleable.MaterialSwitchPreference_preferenceTag);
        mNeedReload = a.getBoolean(R.styleable.MaterialSwitchPreference_needReload, false);
        a.recycle();

        // Handle checkbox
        setChecked(CommonUtils.isIntStrOne(SettingsUtils.getPrefNum(pref, mPreferenceTag)));
        setOnPreferenceClickListener(preference -> {
            SettingsUtils.setPrefIntBoolAccBool(pref, mPreferenceTag, isChecked(), false);
            SettingsActivity.SettingsPrefHandler.needReload = mNeedReload;
            return true;
        });

        // Use material switch
        setWidgetLayoutResource(R.layout.preference_material_switch);
    }
}

File: app/src/main/java/tipz/viola/settings/MaterialSwitchPreference.kt
// Copyright (c) 2023-2024 Tipz Team
// SPDX-License-Identifier: Apache-2.0

package tipz.viola.settings

import android.content.Context
import android.os.Build
import android.util.AttributeSet
import androidx.preference.Preference.OnPreferenceClickListener
import androidx.preference.SwitchPreferenceCompat
import tipz.viola.Application
import tipz.viola.R

class MaterialSwitchPreference(context: Context, attrs: AttributeSet?) :
    SwitchPreferenceCompat(context, attrs) {
    private val settingsPreference: SettingsSharedPreference =
        (context.applicationContext as Application).settingsPreference
    private val mPreferenceTag: String?
    private val mNeedReload: Boolean
    private val mRequiredApi: Int

    init {
        // Get attrs
        val a = context.obtainStyledAttributes(attrs, R.styleable.MaterialSwitchPreference)
        mPreferenceTag = a.getString(R.styleable.MaterialSwitchPreference_preferenceTag)
        mNeedReload = a.getBoolean(R.styleable.MaterialSwitchPreference_needReload, false)
        mRequiredApi = a.getInteger(R.styleable.MaterialSwitchPreference_requiredApi, 1)
        a.recycle()

        // Handle checkbox
        isChecked = settingsPreference.getIntBool(mPreferenceTag!!)
        onPreferenceClickListener = OnPreferenceClickListener {
            settingsPreference.setIntBool(mPreferenceTag, isChecked)
            SettingsActivity.SettingsPrefHandler.needReload = mNeedReload
            true
        }
        if (mRequiredApi > Build.VERSION.SDK_INT) this.isEnabled = false

        // Use material switch
        widgetLayoutResource = R.layout.preference_material_switch
    }
}